instrumentation/undertow-1.4/src/main/java/io/opentelemetry/javaagent/instrumentation/hypertrace/undertow/v1_4/StreamSourceChannelInstrumentation.java

```diff
@@ -0,0 +1,105 @@
+import io.opentelemetry.javaagent.tooling.bytebuddy.matcher.AgentElementMatchers;
+import io.undertow.server.HttpServerExchange;
+import java.nio.ByteBuffer;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.Map;
+import net.bytebuddy.asm.Advice;
+import net.bytebuddy.description.method.MethodDescription;
+import net.bytebuddy.description.type.TypeDescription;
+import net.bytebuddy.matcher.ElementMatcher;
+import net.bytebuddy.matcher.ElementMatcher.Junction;
+import org.hypertrace.agent.core.instrumentation.HypertraceCallDepthThreadLocalMap;
+import org.hypertrace.agent.core.instrumentation.SpanAndBuffer;
+import org.xnio.channels.StreamSourceChannel;
+
+/** Instrumentation for {@link StreamSourceChannel} implementations */
+public final class StreamSourceChannelInstrumentation implements TypeInstrumentation {
+
+  @Override
+  public ElementMatcher<TypeDescription> typeMatcher() {
+    return AgentElementMatchers.safeHasSuperType(named("org.xnio.channels.StreamSourceChannel"));
+  }
+
+  @Override
+  public Map<? extends ElementMatcher<? super MethodDescription>, String> transformers() {
+    final Map<Junction<MethodDescription>, String> matchers = new HashMap<>();
+    matchers.put(
+        named("read")
+            .and(takesArguments(1))
+            .and(takesArgument(0, ByteBuffer.class))
+            .and(returns(int.class))
+            .and(isPublic()),
+        Read_advice.class.getName());
+    return Collections.unmodifiableMap(matchers);
+  }
+
+  /**
+   * Decorates the {@link StreamSourceChannel#read(ByteBuffer)} implementations with logic to
+   * capture data read into the request body {@link ByteBuffer} and report it. This instrumentation
+   * short-circuits if:
+   *
+   * <ul>
+   *   <li>We're in a nested {@link StreamSourceChannel#read(ByteBuffer)} call
+   *   <li>A {@link Throwable} was thrown in the context of the {@link
+   *       StreamSourceChannel#read(ByteBuffer)}, causing the method to exit
+   *   <li>The instrumented {@link StreamSourceChannel} was never put in the {@link
+   *       InstrumentationContext} by {@link
+   *       UndertowHttpServerExchangeInstrumentation.GetRequestChannel_advice#exit(HttpServerExchange,
+   *       StreamSourceChannel)}
+   * </ul>
+   */
+  static final class Read_advice {
+
+    @Advice.OnMethodEnter
+    public static void trackCallDepth() {
+      HypertraceCallDepthThreadLocalMap.incrementCallDepth(StreamSourceChannel.class);
+    }
+
+    @Advice.OnMethodExit(suppress = Throwable.class, onThrowable = Throwable.class)
+    public static void exit(
+        @Advice.Return final int numBytesRead,
+        @Advice.This final StreamSourceChannel streamSourceChannel,
+        @Advice.Thrown final Throwable thrown,
+        @Advice.Argument(0) final ByteBuffer byteBuffer) {
+      if (HypertraceCallDepthThreadLocalMap.decrementCallDepth(StreamSourceChannel.class) > 0
+          || thrown != null) {
```

Comment:
if an exception happened then the counter is not decremented and this advice will never be again executed